File: Valid Palindrome/Valid_Palindrome.cpp
// 题目虽然比较简单，但是需要考虑的条件还不少!
// Runtime: 12 ms, faster than 23.38% of C++ online submissions for Valid Palindrome.
// Memory Usage: 5.4 MB, less than 0.99% of C++ online submissions for Valid Palindrome.

class Solution 
{
public:
    bool isPalindrome(string s) 
    {
        int leftPtr = 0, rightPtr = s.length() - 1;
        while (leftPtr <= rightPtr)
        {
            while (leftPtr <= rightPtr && !judge(s[leftPtr])) leftPtr++;
            while (leftPtr <= rightPtr && !judge(s[rightPtr])) rightPtr--;
            if (leftPtr > rightPtr) break;
            if ((s[leftPtr] == s[rightPtr]) || iscase(s[leftPtr], s[rightPtr]))
                leftPtr++, rightPtr--;
            else
                return false;
        }
        return true;
    }

private:
    inline bool judge(char letter)
    {
        return (letter >= 'a' && letter <= 'z') || 
               (letter >= 'A' && letter <= 'Z') || 
               (letter >= '0' && letter <= '9');
    }

    inline bool iscase(char letter1, char letter2)
    {
        return (letter1 < '0' || letter1 > '9') &&
               (letter2 < '0' || letter2 > '9') &&
               (abs(letter2 - letter1) == 32);
    }
};

File: Valid Palindrome/Valid_Palindrome.py
# Runtime: 76 ms, faster than 47.31% of Python3 online submissions for Valid Palindrome.
# Memory Usage: 6.7 MB, less than 79.51% of Python3 online submissions for Valid Palindrome.
class Solution:
    def isPalindrome(self, s: 'str') -> 'bool':
        leftPtr, rightPtr = 0, len(s) - 1
        while (leftPtr <= rightPtr):
            while (leftPtr <= rightPtr and not self.judge(s[leftPtr])):
                leftPtr += 1
            while (leftPtr <= rightPtr and not self.judge(s[rightPtr])):
                rightPtr -= 1
            if (leftPtr > rightPtr):
                break
            if (s[leftPtr] == s[rightPtr] or self.iscase(s[leftPtr], s[rightPtr])):
                leftPtr += 1
                rightPtr -= 1
            else:
                return False
        return True
    
    def judge(self, letter):
        return (('a' <= letter <= 'z') or 
                ('A' <= letter <= 'Z') or 
                ('0' <= letter <= '9'))
    
    def iscase(self, letter1, letter2):
        return ((letter1 < '0' or letter1 > '9') and 
                (letter2 < '0' or letter2 > '9') and
                (abs(ord(letter2) - ord(letter1)) is 32))
